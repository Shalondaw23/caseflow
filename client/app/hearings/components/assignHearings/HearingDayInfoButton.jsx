import PropTypes from 'prop-types';
import React from 'react';
import moment from 'moment';

import Tooltip from '../../../components/Tooltip';
import Button from '../../../components/Button';
import {
  buttonSelectedStyle,
  buttonUnselectedStyle,
  dateStyle,
  leftColumnStyle,
  rightColumnStyle,
  typeAndJudgeStyle,
  slotDisplayStyle
} from './styles';

import {
  formatHearingType,
  vljFullnameOrEmptyString,
  formatSlotRatio,
  hearingDayHasJudge,
  separatorIfJudgeOrRoomPresent,
  hearingRoomOrEmptyString
} from '../../utils';

export const HearingDayInfoButton = ({ id, hearingDay, selected, onSelectedHearingDayChange }) => {
  // Format the pieces of information from hearingDay
  const formattedHearingType = formatHearingType(hearingDay.readableRequestType);
  const judgeOrRoom = hearingDayHasJudge(hearingDay) ?
    vljFullnameOrEmptyString(hearingDay) :
    hearingRoomOrEmptyString(hearingDay);
  const separator = separatorIfJudgeOrRoomPresent(hearingDay);
  const formattedSlotRatio = formatSlotRatio(hearingDay);
  const formattedDate = moment(hearingDay.scheduledFor).format('ddd MMM D');

  return (
    <Tooltip id={`hearing-day-${id}`} text={judgeOrRoom} position="bottom" tabIndex={-1}>
      <Button
        styling={selected ? buttonSelectedStyle : buttonUnselectedStyle}
        onClick={() => onSelectedHearingDayChange(hearingDay)}
        // classNames={classNames}
        linkStyling>

        <div>
          <div {...leftColumnStyle} >
            <div {...dateStyle}>
              {formattedDate}
            </div>
            <div {...typeAndJudgeStyle}>
              {`${formattedHearingType} ${separator} ${judgeOrRoom}`}
            </div>
          </div>
          <div {...rightColumnStyle} >
            <div {...slotDisplayStyle}>
              {formattedSlotRatio}
            </div>
            <div>scheduled</div>
          </div>
        </div>
      </Button>
    </Tooltip>
  );
};

HearingDayInfoButton.propTypes = {
  id: PropTypes.number,
  hearingDay: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.object
  ]),
  selected: PropTypes.bool,
  onSelectedHearingDayChange: PropTypes.func,
};

export default HearingDayInfoButton;
