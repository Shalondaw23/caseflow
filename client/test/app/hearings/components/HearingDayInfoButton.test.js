import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { axe } from 'jest-axe';

import HearingDayInfoButton from 'app/hearings/components/assignHearings/HearingDayInfoButton';

describe('HearingDayInfoButton', () => {
  const setSelectedValue = jest.fn();

  const props = {
    hearingDay:
      {
        judgeFirstName: 'Jonas',
        judgeLastName: 'Jerengal',
        room: '14',
        readableRequestType: 'video',
        hearings: {
          hearing_1: {},
          hearing_2: {},
        },
        totalSlots: 8,
      },
    selected: false,
    onSelectedHearingDayChange: setSelectedValue,
  };

  it('renders correctly', () => {
    const { container } = render(<HearingDayInfoButton {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('passes a11y testing', async () => {
    const { container } = render(<HearingDayInfoButton {...props} />);

    const results = await axe(container);

    expect(results).toHaveNoViolations();
  });

  /*
  it('correctly sets checkbox status when all are unchecked', () => {
    const uncheckedOptions = props.options.map((opt) => {
      return { ...opt, checked: false };
    });
    const allFalseOptionsProps = { ...props, options: uncheckedOptions };

    const component = render(<FilterOption {...allFalseOptionsProps} />);
    const options = component.getAllByRole('checkbox');
    const checked = options.filter((el) => el.checked);

    expect(checked.length).toBe(0);
  });

  it('correctly sets checkbox status when all are checked', () => {
    const checkedOptions = props.options.map((opt) => {
      return { ...opt, checked: true };
    });
    const allTrueOptionsProps = { ...props, options: checkedOptions };

    const component = render(<FilterOption {...allTrueOptionsProps} />);
    const options = component.getAllByRole('checkbox');
    const checked = options.filter((el) => el.checked);

    expect(checked.length).toBe(options.length);
  });

  it('correctly calls setSelectedValue', async () => {
    const component = render(<FilterOption {...props} />);

    const options = component.getAllByRole('checkbox');
    const opt = props.options[1];
    const checkedBefore = options.filter((el) => el.checked);

    expect(checkedBefore.length).toBe(1);

    await userEvent.click(screen.getByLabelText(opt.displayText));

    // Calls onChange handler
    expect(setSelectedValue).toHaveBeenCalledTimes(1);

    // While we aren't updating value, handleChange is still getting called
    expect(setSelectedValue).toHaveBeenLastCalledWith(opt.value);

    const checkedAfter = options.filter((el) => el.checked);

    // Value should not have been updated, since we're not doing that
    expect(checkedAfter.length).toBe(1);
  });
  */
});
