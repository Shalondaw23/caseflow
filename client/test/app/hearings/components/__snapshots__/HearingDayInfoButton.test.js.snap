// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HearingDayInfoButton renders correctly 1`] = `
<div>
  <span>
    <button
      class="unselected-hearing-day-info-button cf-btn-link usa-button"
      data-css-s0bvcw=""
      type="button"
    >
      <div>
        <div
          data-css-16v9b8i=""
        >
          <div
            data-css-l6c9n4=""
          >
            Mon Apr 12th
          </div>
          <div
            data-css-134a1o=""
          >
            Video Â· Jerengal, Jonas
          </div>
        </div>
        <div
          data-css-1fnaqki=""
        >
          <div>
            2 of 6
          </div>
          <div>
            scheduled
          </div>
        </div>
      </div>
    </button>
  </span>
</div>
`;
